File: control/python_interface/exercise_3/exercise_3.py
#!/usr/bin/env python3

# Copyright 2026 Universidad Politécnica de Madrid
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
"""Drone Controller Node.

This module implements a ROS2 node for controlling a drone using the
Aerostack2 framework. It handles arming, offboard mode, control mode
setting, and position commands.
"""

import future
from as2_msgs.msg import ControlMode
from as2_msgs.srv import SetControlMode
from drone_course_msgs.srv import RequestPath
from geometry_msgs.msg import PoseStamped, TwistStamped
import rclpy
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.node import Node
from rclpy.qos import HistoryPolicy, QoSProfile, ReliabilityPolicy


class DroneController(Node):
    """Drone Controller class for managing drone operations.

    This class handles subscribing to drone state information,
    publishing motion commands, and calling services to control the
    drone's operational state.
    """

    def __init__(self):
        """Initialize the DroneController node.

        Sets up subscribers, publishers, service clients, and the main
        control timer.
        """
        super().__init__('drone_controller')
        # QoS profile for reliable topics
        reliable_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE, history=HistoryPolicy.KEEP_LAST, depth=10
        )

        # QoS profile for best effort topics
        best_effort_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT, history=HistoryPolicy.KEEP_LAST, depth=10
        )
        
        self.callback_group = ReentrantCallbackGroup()

        # Subscribers
        
        self.pose_sub = self.create_subscription(
            PoseStamped,
            '/drone0/self_localization/pose',
            self.state_pose_callback,
            best_effort_qos,
        )

        # Publishers

        self.twist_pub = self.create_publisher(
            TwistStamped, '/drone0/motion_reference/twist', reliable_qos
        )

        # Service clients

        self.control_mode_client = self.create_client(
            SetControlMode, '/drone0/controller/set_control_mode', callback_group=self.callback_group
        )

        self.path_request_client = self.create_client(RequestPath, '/request_path', callback_group=self.callback_group)
        while not self.path_request_client.service_is_ready():
            self.get_logger().warn('Path request service is not available')

        # Class variables
        self.state_pose: PoseStamped = PoseStamped()
        self.control_mode_set: bool = False
        self.path: list[float] = []
        self.path_idx: int = 0

        # Timer at 100Hz (0.01 seconds)
        self.timer = self.create_timer(0.01, self.run, callback_group=self.callback_group)

        self.get_logger().info('Drone Controller Node initialized')

    def state_pose_callback(self, msg):
        """Callback for state pose messages.

        :param msg: The received pose message
        :type msg: PoseStamped
        """
        self.state_pose = msg

    def run(self):
        """Main control loop.

        Implements the state machine for drone control:
        1. Arm the drone
        2. Set offboard mode
        3. Set control mode to position
        4. Send motion reference commands
        5. Monitor drone state
        """
        
        # Request path from service
        if not len(self.path) > 0:
            req = RequestPath.Request()
            response : RequestPath.Response = self.path_request_client.call(req)
            self.path = response.path
        
        # Check if control mode is set, if not, call the service to set it
        if not self.control_mode_set:
            self.get_logger().info('Calling control mode service...')

            # Check if service is available
            if not self.control_mode_client.service_is_ready():
                self.get_logger().warn('Control mode service is not available')
                return

            # Generate control mode service request
            request = SetControlMode.Request()
            control_mode: ControlMode = ControlMode()
            control_mode.control_mode = ControlMode.SPEED
            control_mode.yaw_mode = ControlMode.YAW_SPEED
            control_mode.reference_frame = ControlMode.LOCAL_ENU_FRAME
            request.control_mode = control_mode

            response: SetControlMode.Response = self.control_mode_client.call(request)
            success = response.success
            if not success:
                self.get_logger().warn('Failed to set control mode')
                return
            self.get_logger().info('Control mode set successfully')
            self.control_mode_set = True

        # TODO (Exercise 3):
        # Modify reference generation to follow a smooth trajectory
        position_ref = self.path[self.path_idx]
        position_ref = (position_ref.x, position_ref.y, position_ref.z)

        # Read current drone state
        current_x = self.state_pose.pose.position.x
        current_y = self.state_pose.pose.position.y
        current_z = self.state_pose.pose.position.z

        position_error_x = position_ref[0] - current_x
        position_error_y = position_ref[1] - current_y
        position_error_z = position_ref[2] - current_z
        position_error = (position_error_x**2 + position_error_y**2 + position_error_z**2) ** 0.5

        Kp = 1.0
        
        velocity_command_x = Kp * position_error_x
        velocity_command_y = Kp * position_error_y
        velocity_command_z = Kp * position_error_z    
        
        twist_msg = TwistStamped()
        twist_msg.header.stamp = self.get_clock().now().to_msg()
        twist_msg.header.frame_id = 'earth'
        twist_msg.twist.linear.x = velocity_command_x
        twist_msg.twist.linear.y = velocity_command_y
        twist_msg.twist.linear.z = velocity_command_z
        twist_msg.twist.angular.x = 0.0
        twist_msg.twist.angular.y = 0.0
        twist_msg.twist.angular.z = 0.0
        self.twist_pub.publish(twist_msg)

        if position_error < 0.2:
            self.get_logger().info(
                'Drone has reached the target position', throttle_duration_sec=1.0
            )
            self.path_idx = (self.path_idx + 1) % len(self.path)


def main(args=None):
    """Main entry point for the drone controller node.

    :param args: Command line arguments
    :type args: list
    """
    rclpy.init(args=args)
    drone_controller = DroneController()

    # Use MultiThreadedExecutor to allow synchronous service calls from timer
    from rclpy.executors import MultiThreadedExecutor

    executor = MultiThreadedExecutor()
    executor.add_node(drone_controller)

    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        drone_controller.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: control/drone_course_ws/src/exercise_3/src/exercise_3.cpp
// Copyright 2026 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.


#include <exercise_3/exercise_3.hpp>

namespace drone_course
{

DroneCourseExercise3::DroneCourseExercise3(
  const std::string & node_name,
  const rclcpp::NodeOptions & options)
: rclcpp::Node(node_name, options)
{
  // QoS profiles
  rclcpp::QoS reliable_qos = rclcpp::QoS(10).reliable();
  rclcpp::QoS best_effort_qos = rclcpp::QoS(10).best_effort();

  callback_group_ =
    this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);

  // Suscribers

  pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
    "/drone0/self_localization/pose", best_effort_qos,
    std::bind(&DroneCourseExercise3::state_subscription_callback, this, std::placeholders::_1));

  // Publishers

  vel_pub_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(
    "/drone0/motion_reference/twist", reliable_qos);

  // Services clients

  control_mode_service_client_ = this->create_client<as2_msgs::srv::SetControlMode>(
    "/drone0/controller/set_control_mode", rmw_qos_profile_services_default, callback_group_);

  path_service_client_ = this->create_client<drone_course_msgs::srv::RequestPath>(
    "/request_path", rmw_qos_profile_services_default, callback_group_);

  // Timers
  double timer_freq = 100.0;
  timer_ = this->create_wall_timer(
    std::chrono::duration<double>(1.0f / timer_freq),
    std::bind(&DroneCourseExercise3::timer_callback, this));

  dt_ = 1.0f / timer_freq;

  RCLCPP_INFO(this->get_logger(), "DroneCourseExercise2 initialized\n");
}

DroneCourseExercise3::~DroneCourseExercise3() {}

void DroneCourseExercise3::control_mode_service_callback(
  rclcpp::Client<as2_msgs::srv::SetControlMode>::SharedFuture future)
{
  auto status = future.wait_for(std::chrono::seconds(1));
  if (status == std::future_status::ready) {
    as2_msgs::srv::SetControlMode::Response::SharedPtr response = future.get();
    if (response->success) {
      RCLCPP_INFO(this->get_logger(), "Control mode set successfully");
      control_mode_set_ = true;
    } else {
      RCLCPP_ERROR(this->get_logger(), "Failed to set control mode");
    }
  } else {
    RCLCPP_INFO(this->get_logger(), "Service In-Progress...");
  }
}

void DroneCourseExercise3::path_service_callback(
  rclcpp::Client<drone_course_msgs::srv::RequestPath>::SharedFuture future)
{
  auto status = future.wait_for(std::chrono::seconds(1));
  if (status == std::future_status::ready) {
    drone_course_msgs::srv::RequestPath::Response::SharedPtr response = future.get();
    if (response->path.size() > 0) {
      RCLCPP_INFO(this->get_logger(), "Received path from service");
      path_ = response->path;
      path_received_ = true;
    } else {
      RCLCPP_INFO(this->get_logger(), "Could not retrieve path from service");
    }
  } else {
    RCLCPP_INFO(this->get_logger(), "Service In-Progress...");
  }
}


void DroneCourseExercise3::timer_callback()
{
  if (!path_received_) {
    RCLCPP_INFO(this->get_logger(), "Requesting path from service...");
    path_service_request_ = std::make_shared<drone_course_msgs::srv::RequestPath::Request>();

    path_service_client_->async_send_request(
      path_service_request_, std::bind(
        &DroneCourseExercise3::path_service_callback, this, std::placeholders::_1));

  }

  // Check if control mode is set, if not, call service
  if (!control_mode_set_) {
    RCLCPP_INFO(this->get_logger(), "Calling control mode service...");
    as2_msgs::srv::SetControlMode::Request::SharedPtr control_mode_request =
      std::make_shared<as2_msgs::srv::SetControlMode::Request>();
    control_mode_request->control_mode.control_mode = as2_msgs::msg::ControlMode::SPEED;
    control_mode_request->control_mode.yaw_mode = as2_msgs::msg::ControlMode::YAW_SPEED;
    control_mode_request->control_mode.reference_frame =
      as2_msgs::msg::ControlMode::LOCAL_ENU_FRAME;

    control_mode_service_client_->async_send_request(
      control_mode_request, std::bind(
        &DroneCourseExercise3::control_mode_service_callback, this, std::placeholders::_1));
  }

  // TODO (Exercise 3):
  // Modify reference generation to follow a smooth trajectory
  // Desired position reference
  drone_course_msgs::msg::Point position_ref;
  if (path_.size() > 0) {
    position_ref = path_[path_index_];
  } else {
    position_ref.x = 0.0;
    position_ref.y = 0.0;
    position_ref.z = 1.0;
  }


  // Read current drone state
  double current_x = state_pose_.pose.position.x;
  double current_y = state_pose_.pose.position.y;
  double current_z = state_pose_.pose.position.z;

  double position_error_x = position_ref.x - current_x;
  double position_error_y = position_ref.y - current_y;
  double position_error_z = position_ref.z - current_z;
  double position_error_norm = std::sqrt(
    position_error_x * position_error_x + position_error_y * position_error_y + position_error_z *
    position_error_z);

  // Tune PID controller gains
  double Kp = 1.0;

  // Compute PID control action
  double velocity_x = Kp * position_error_x;
  double velocity_y = Kp * position_error_y;
  double velocity_z = Kp * position_error_z;

  // Store current error for next iteration
  prev_error_[0] = position_error_x;
  prev_error_[1] = position_error_y;
  prev_error_[2] = position_error_z;control / python_interface / exercise_2

  // Generate motion reference command
  geometry_msgs::msg::TwistStamped velocity_msg;
  velocity_msg.header.stamp = this->get_clock()->now();
  velocity_msg.header.frame_id = "earth";
  velocity_msg.twist.linear.x = velocity_x;
  velocity_msg.twist.linear.y = velocity_y;
  velocity_msg.twist.linear.z = velocity_z;
  velocity_msg.twist.angular.x = 0.0;
  velocity_msg.twist.angular.y = 0.0;
  velocity_msg.twist.angular.z = 0.0;

  if (vel_pub_) {
    vel_pub_->publish(velocity_msg);
  }

  if (position_error_norm < 0.2) {
    RCLCPP_INFO(this->get_logger(), "Drone has reach the target position");
    if (path_.size() > 0) {
      path_index_ = (path_index_ + 1) % (path_.size());
    }
  }
}

void DroneCourseExercise3::state_subscription_callback(
  const geometry_msgs::msg::PoseStamped::SharedPtr msg)
{
  state_pose_ = *msg;
}
}  // namespace drone_course
